File: macos/NearClip/Sources/NearClip/NetworkMonitor.swift
import Foundation
import Network

/// Monitors network connectivity and triggers reconnection on recovery
final class NetworkMonitor {
    static let shared = NetworkMonitor()

    // MARK: - Constants

    /// Maximum number of reconnection attempts before giving up
    private static let maxReconnectAttempts = 3

    /// Base delay for exponential backoff (seconds)
    private static let baseReconnectDelay: TimeInterval = 1.0

    /// Delay to verify connection status after reconnection attempt (seconds)
    private static let connectionVerifyDelay: TimeInterval = 5.0

    // MARK: - Properties

    private var monitor: NWPathMonitor?
    private let queue = DispatchQueue(label: "com.nearclip.networkmonitor")

    private var isConnected = false
    private var wasDisconnected = false
    private var reconnectAttempts = 0
    private var reconnectWorkItem: DispatchWorkItem?
    private var isMonitoring = false

    /// Callback when network connectivity is restored
    var onNetworkRestored: (() -> Void)?

    /// Callback when network connectivity is lost
    var onNetworkLost: (() -> Void)?

    /// Callback when reconnection fails after max attempts
    var onReconnectFailed: (() -> Void)?

    private init() {}

    // MARK: - Public API

    /// Start monitoring network connectivity
    func startMonitoring() {
        guard !isMonitoring else {
            print("NetworkMonitor: Already monitoring")
            return
        }

        // Create a new monitor instance (NWPathMonitor cannot be restarted after cancel)
        let newMonitor = NWPathMonitor()
        newMonitor.pathUpdateHandler = { [weak self] path in
            self?.handlePathUpdate(path)
        }
        newMonitor.start(queue: queue)
        monitor = newMonitor
        isMonitoring = true
        print("NetworkMonitor: Started monitoring")
    }

    /// Stop monitoring network connectivity
    func stopMonitoring() {
        guard isMonitoring else { return }

        monitor?.cancel()
        monitor = nil
        reconnectWorkItem?.cancel()
        reconnectWorkItem = nil
        isMonitoring = false
        print("NetworkMonitor: Stopped monitoring")
    }

    /// Reset reconnection attempts counter
    func resetReconnectAttempts() {
        reconnectAttempts = 0
    }

    // MARK: - Private

    private func handlePathUpdate(_ path: NWPath) {
        let newIsConnected = path.status == .satisfied

        print("NetworkMonitor: Path update - status: \(path.status), isExpensive: \(path.isExpensive), isConstrained: \(path.isConstrained)")

        if !newIsConnected && isConnected {
            // Network went down
            wasDisconnected = true
            print("NetworkMonitor: Network disconnected")

            // Notify about network loss (for BLE fallback)
            DispatchQueue.main.async { [weak self] in
                self?.onNetworkLost?()
            }
        } else if newIsConnected && wasDisconnected {
            // Network recovered
            wasDisconnected = false
            print("NetworkMonitor: Network restored, scheduling reconnection")
            scheduleReconnect()
        }

        isConnected = newIsConnected
    }

    private func scheduleReconnect() {
        // Cancel any pending reconnect
        reconnectWorkItem?.cancel()

        // Delay reconnection to let network stabilize
        let delay = calculateReconnectDelay()

        let workItem = DispatchWorkItem { [weak self] in
            self?.attemptReconnect()
        }
        reconnectWorkItem = workItem

        DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: workItem)
        print("NetworkMonitor: Reconnection scheduled in \(delay)s (attempt \(reconnectAttempts + 1)/\(Self.maxReconnectAttempts))")
    }

    private func calculateReconnectDelay() -> TimeInterval {
        // Exponential backoff: 1s, 2s, 4s
        return Self.baseReconnectDelay * pow(2.0, Double(reconnectAttempts))
    }

    private func attemptReconnect() {
        reconnectAttempts += 1

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            print("NetworkMonitor: Attempting reconnection (attempt \(self.reconnectAttempts)/\(Self.maxReconnectAttempts))")

            // Trigger reconnection
            self.onNetworkRestored?()

            // Check if we need to retry after a delay to allow connection to establish
            DispatchQueue.main.asyncAfter(deadline: .now() + Self.connectionVerifyDelay) { [weak self] in
                guard let self = self else { return }

                // If still not connected after callback, and we haven't exceeded max attempts
                let connectionManager = ConnectionManager.shared
                if connectionManager.connectedDevices.isEmpty && self.reconnectAttempts < Self.maxReconnectAttempts {
                    // Schedule another attempt
                    self.scheduleReconnect()
                } else if connectionManager.connectedDevices.isEmpty && self.reconnectAttempts >= Self.maxReconnectAttempts {
                    // Max attempts reached, notify user
                    print("NetworkMonitor: Reconnection failed after \(Self.maxReconnectAttempts) attempts")
                    self.onReconnectFailed?()
                } else {
                    // Successfully reconnected
                    print("NetworkMonitor: Reconnection successful")
                    self.reconnectAttempts = 0
                }
            }
        }
    }
}

File: android/app/src/main/java/com/nearclip/service/NetworkMonitor.kt
package com.nearclip.service

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Handler
import android.os.Looper
import android.util.Log

/**
 * Monitors network connectivity and triggers reconnection on recovery.
 */
class NetworkMonitor(private val context: Context) {

    companion object {
        private const val TAG = "NetworkMonitor"
        private const val MAX_RECONNECT_ATTEMPTS = 3
        private const val BASE_RECONNECT_DELAY_MS = 1000L
    }

    private val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    private val handler = Handler(Looper.getMainLooper())

    private var isRegistered = false
    private var wasDisconnected = false
    private var reconnectAttempts = 0
    private var pendingReconnect: Runnable? = null

    /** Callback when network connectivity is restored */
    var onNetworkRestored: (() -> Unit)? = null

    /** Callback when reconnection fails after max attempts */
    var onReconnectFailed: (() -> Unit)? = null

    /** Callback to check if currently connected to devices */
    var isConnectedToDevices: (() -> Boolean)? = null

    private val networkCallback = object : ConnectivityManager.NetworkCallback() {
        override fun onAvailable(network: Network) {
            Log.d(TAG, "Network available")
            if (wasDisconnected) {
                wasDisconnected = false
                Log.i(TAG, "Network restored, scheduling reconnection")
                scheduleReconnect()
            }
        }

        override fun onLost(network: Network) {
            Log.d(TAG, "Network lost")
            wasDisconnected = true
        }

        override fun onCapabilitiesChanged(network: Network, networkCapabilities: NetworkCapabilities) {
            val hasInternet = networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            val hasValidated = networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
            Log.d(TAG, "Network capabilities changed: hasInternet=$hasInternet, hasValidated=$hasValidated")
        }
    }

    /**
     * Start monitoring network connectivity.
     */
    fun startMonitoring() {
        if (isRegistered) return

        val request = NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .build()

        try {
            connectivityManager.registerNetworkCallback(request, networkCallback)
            isRegistered = true
            Log.i(TAG, "Started monitoring network connectivity")

            // Check current network state
            val currentNetwork = connectivityManager.activeNetwork
            val capabilities = currentNetwork?.let { connectivityManager.getNetworkCapabilities(it) }
            val hasInternet = capabilities?.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) == true
            wasDisconnected = !hasInternet

        } catch (e: Exception) {
            Log.e(TAG, "Failed to register network callback", e)
        }
    }

    /**
     * Stop monitoring network connectivity.
     */
    fun stopMonitoring() {
        if (!isRegistered) return

        try {
            connectivityManager.unregisterNetworkCallback(networkCallback)
            isRegistered = false
            cancelPendingReconnect()
            Log.i(TAG, "Stopped monitoring network connectivity")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to unregister network callback", e)
        }
    }

    /**
     * Reset reconnection attempts counter.
     */
    fun resetReconnectAttempts() {
        reconnectAttempts = 0
    }

    private fun scheduleReconnect() {
        cancelPendingReconnect()

        val delay = calculateReconnectDelay()
        Log.i(TAG, "Reconnection scheduled in ${delay}ms (attempt ${reconnectAttempts + 1}/$MAX_RECONNECT_ATTEMPTS)")

        val runnable = Runnable {
            attemptReconnect()
        }
        pendingReconnect = runnable
        handler.postDelayed(runnable, delay)
    }

    private fun cancelPendingReconnect() {
        pendingReconnect?.let { handler.removeCallbacks(it) }
        pendingReconnect = null
    }

    private fun calculateReconnectDelay(): Long {
        // Exponential backoff: 1s, 2s, 4s
        return BASE_RECONNECT_DELAY_MS * (1L shl reconnectAttempts)
    }

    private fun attemptReconnect() {
        reconnectAttempts++
        Log.i(TAG, "Attempting reconnection (attempt $reconnectAttempts/$MAX_RECONNECT_ATTEMPTS)")

        // Trigger reconnection
        onNetworkRestored?.invoke()

        // Check result after a delay
        handler.postDelayed({
            val isConnected = isConnectedToDevices?.invoke() ?: false

            if (!isConnected && reconnectAttempts < MAX_RECONNECT_ATTEMPTS) {
                // Schedule another attempt
                scheduleReconnect()
            } else if (!isConnected && reconnectAttempts >= MAX_RECONNECT_ATTEMPTS) {
                // Max attempts reached, notify user
                Log.w(TAG, "Reconnection failed after $MAX_RECONNECT_ATTEMPTS attempts")
                onReconnectFailed?.invoke()
            } else {
                // Successfully reconnected
                Log.i(TAG, "Reconnection successful")
                reconnectAttempts = 0
            }
        }, 5000L)
    }
}
